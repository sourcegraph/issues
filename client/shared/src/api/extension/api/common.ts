import { Remote, ProxyMarked, proxy, proxyMarker, UnproxyOrClone } from 'comlink'
import { identity } from 'lodash'
import { from, isObservable, Observable, Observer, of } from 'rxjs'
import { map } from 'rxjs/operators'
import { ProviderResult, Subscribable, Unsubscribable } from 'sourcegraph'

import { isAsyncIterable, isPromiseLike, isSubscribable, observableFromAsyncIterable } from '../../util'

/**
 * A Subscribable that can be exposed by comlink to the other thread.
 * Only allows full object Observers to avoid complex type checking against proxies.
 */
export interface ProxySubscribable<T> extends ProxyMarked {
    subscribe(observer: Remote<Observer<T> & ProxyMarked>): Unsubscribable & ProxyMarked
}

/**
 * Wraps a given Subscribable so that it is exposed by comlink to the other thread.
 *
 * @param subscribable A normal Subscribable (from this thread)
 */
export const proxySubscribable = <T>(subscribable: Subscribable<T>): ProxySubscribable<T> => ({
    [proxyMarker]: true,
    subscribe(observer): Unsubscribable & ProxyMarked {
        return proxy(
            // Don't pass the proxy to Rx directly because it will try to
            // access Symbol properties that cannot be proxied
            subscribable.subscribe({
                next: value => {
                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
                    observer.next(value as UnproxyOrClone<T>)
                },
                error: error => {
                    // Only pass a few well-known Error properties
                    // TODO should pass all properties serialized recursively, best handled on comlink level
                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
                    observer.error(error && { message: error.message, name: error.name, stack: error.stack })
                },
                complete: () => {
                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
                    observer.complete()
                },
            })
        )
    },
})

export function providerResultToObservable<T, R = T>(
    result: ProviderResult<T>,
    mapFunc: (value: T | undefined | null) => R = identity
): Observable<R> {
    let observable: Observable<R>
    if (result && (isPromiseLike(result) || isObservable<T>(result) || isSubscribable(result))) {
        observable = from(result).pipe(map(mapFunc))
    } else if (isAsyncIterable(result)) {
        observable = observableFromAsyncIterable(result).pipe(map(mapFunc))
    } else {
        observable = of(mapFunc(result))
    }
    return observable
}

interface Deferred<Value = unknown> extends Promise<Value> {
    resolve(value?: Value): void;
    reject(error: Error): void;
}

const createAsyncMutex = <Value,>(): Deferred<Value> => {
    const methodStore = {} as Deferred<Value>;
    const promise = new Promise((resolve, reject) => {
        methodStore.resolve = resolve;
        methodStore.reject = resolve;
    })

    return {...promise, ...methodStore};
}

export async function* observableToAsyncGenerator<D>(observable: Observable<D>): AsyncIterableIterator<D> {
    let asyncMutex = createAsyncMutex<D>();
    let completed = false;

    const subscription = observable.subscribe({
        next: value => {
            const currentAsyncMutex = asyncMutex;
            asyncMutex = createAsyncMutex<D>();

            currentAsyncMutex.resolve(value);
        },
        error: error => {
            const currentAsyncMutex = asyncMutex;
            asyncMutex = createAsyncMutex<D>();
            currentAsyncMutex.reject(error instanceof Error ? error : new Error(String(error)));
        },
        complete: () => {
            completed = true;
            asyncMutex.resolve();
        },
    })

    try {
        while(true) {
            const nextValue = await asyncMutex;

            if (completed) {
                break
            }

            yield nextValue;
        }
    } finally {
        subscription.unsubscribe()
    }
}

