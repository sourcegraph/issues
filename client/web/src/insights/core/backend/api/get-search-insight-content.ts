import { formatISO, isAfter, startOfDay, sub } from 'date-fns';
import escapeRegExp from 'lodash/escapeRegExp'
import { defer } from 'rxjs';
import { retry } from 'rxjs/operators';
import sourcegraph from 'sourcegraph';

import { ICommitSearchResult } from '@sourcegraph/shared/src/graphql/schema';

import { fetchRawSearchInsightResults, fetchSearchInsightCommits } from '../requests/fetch-search-insight.ignored';
import { SearchInsight } from '../types';

/**
 * This logic is a copy of fetch logic of search-based code insight extension.
 * In order to have live preview for creation UI we had to copy this logic from
 * extension.
 *
 * */
export async function getSearchInsightContent(insight: SearchInsight): Promise<sourcegraph.View> {
    const step = insight.step || { days: 1 }
    const { repositories: repos } = insight
    const dates = getDaysToQuery(step)

    // Get commits to search for each day
    const repoCommits = (
        await Promise.all(
            repos.map(async repo => (await determineCommitsToSearch(dates, repo)).map(commit => ({ repo, ...commit })))
        )
    ).flat()

    const searchQueries = insight.series.flatMap(({ query, name }) =>
        repoCommits.map(({ date, repo, commit }) => ({
            name,
            date,
            repo,
            commit,
            query: `repo:^${escapeRegExp(repo)}$@${commit} ${query} 'count:99999'`,
        }))
    )
    const rawSearchResults = await defer(() =>
        fetchRawSearchInsightResults(searchQueries.map(search => search.query)))
        // The bulk search may timeout, but a retry is then likely faster because caches are warm
        .pipe(retry(3))
        .toPromise()

    const searchResults = Object.entries(rawSearchResults).map(([field, result]) => {
        const index = +field.slice('search'.length)
        const query = searchQueries[index]
        return { ...query, result }
    })

    const data: {
        date: number
        [seriesName: string]: number
    }[] = []
    for (const { name, date, result } of searchResults) {
        const dataKey = name
        const dataIndex = dates.indexOf(date)
        const object =
            data[dataIndex] ??
            (data[dataIndex] = {
                date: date.getTime(),
                // Initialize all series to 0
                ...Object.fromEntries(insight.series.map(series => [series.name, 0])),
            })
        // Sum across repos
        const countForRepo = result?.results.matchCount

        object[dataKey] += countForRepo ?? 0
    }

    return {
        title: insight.title,
        subtitle: insight.subtitle,
        content: [
            {
                chart: 'line' as const,
                data,
                series: insight.series.map(series => ({
                    dataKey: series.name,
                    name: series.name,
                    stroke: series.color,
                    linkURLs: dates.map(date => {
                        // Link to diff search that explains what new cases were added between two data points
                        const url = new URL('/search', sourcegraph.internal.sourcegraphURL)
                        // Use formatISO instead of toISOString(), because toISOString() always outputs UTC.
                        // They mark the same point in time, but using the user's timezone makes the date string
                        // easier to read (else the date component may be off by one day)
                        const after = formatISO(sub(date, step))
                        const before = formatISO(date)
                        const repoFilters = repos.map(repo => `repo:^${escapeRegExp(repo)}$`).join(' ')
                        const diffQuery = `${repoFilters} type:diff after:${after} before:${before} ${series.query}`
                        url.searchParams.set('q', diffQuery)
                        return url.href
                    }),
                })),
                xAxis: {
                    dataKey: 'date' as const,
                    type: 'number' as const,
                    scale: 'time' as const,
                },
            },
        ],
    }
}

async function determineCommitsToSearch(dates: Date[], repo: string): Promise<{ date: Date; commit: string }[]> {
    const commitQueries = dates.map(date => {
        const before = formatISO(date)
        return `repo:^${escapeRegExp(repo)}$ type:commit before:${before} count:1`
    })

    console.log('searching commits for search live preview', commitQueries)

    const commitResults = await fetchSearchInsightCommits(commitQueries).toPromise();

    const commitOids = Object.entries(commitResults).map(([name, search], index) => {
        const index_ = +name.slice('search'.length)

        if (index_ !== index) {
            throw new Error(`Expected field ${name} to be at index ${index_} of object keys`)
        }

        if (search?.results?.results?.length ?? 0 === 0) {
            throw new Error(`No result for ${commitQueries[index_]}`)
        }

        const commit = (search?.results.results[0] as ICommitSearchResult).commit

        // Sanity check
        const commitDate = commit.committer && new Date(commit.committer.date)
        const date = dates[index_]
        if (!commitDate) {
            throw new Error(`Expected commit to have committer: \`${commit.oid}\``)
        }
        if (isAfter(commitDate, date)) {
            throw new Error(
                `Expected commit \`${commit.oid}\` to be before ${formatISO(date)}, but was after: ${formatISO(
                    commitDate
                )}.\nSearch query: ${commitQueries[index_]}`
            )
        }

        return { commit: commit.oid, date }
    })

    return commitOids
}

function getDaysToQuery(step: globalThis.Duration): Date[] {
    const now = startOfDay(new Date())
    const dates: Date[] = []

    for (let index = 0, date = now; index < 7; index++) {
        dates.unshift(date)
        date = sub(date, step)
    }

    return dates
}
