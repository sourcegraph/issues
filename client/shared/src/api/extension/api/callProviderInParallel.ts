import { concat, from, Observable, of } from 'rxjs'
import { catchError, debounceTime, defaultIfEmpty, map, mergeMap, scan, switchMap } from 'rxjs/operators'
import sourcegraph from 'sourcegraph'

import { asError, ErrorLike } from '../../../util/errors'
import { allOf, isDefined, isExactly, isNot, property } from '../../../util/types'
import { ContributableViewContainer } from '../../protocol'
import { RegisteredViewProvider, ViewContexts, ViewProviderResult } from '../extensionHostApi'

import { providerResultToObservable } from './common'

const DEFAULT_MAX_PARALLEL_QUERIES = 2

/** Load view providers in parallel with parallel queries limit. */
export function callViewProvidersInParallel<W extends ContributableViewContainer>(
    context: ViewContexts[W],
    providers: Observable<readonly RegisteredViewProvider<W>[]>,
    maxParallelQueries = DEFAULT_MAX_PARALLEL_QUERIES
): Observable<ViewProviderResult[]> {
    return providers.pipe(
        debounceTime(0),
        switchMap(providers =>
            // Add first synthetic observable with null withing to trigger
            // all operators chain immediately in first time
            concat(of(null), from(providers)).pipe(
                mergeMap(
                    (provider, index) =>
                        provider
                            ? // Just because we have this first nullable synthetic event we have to avoid
                              // calling provideView on null value
                              providerResultToObservable(provider.viewProvider.provideView(context)).pipe(
                                  defaultIfEmpty<sourcegraph.View | null | undefined>(null),
                                  catchError((error): [ErrorLike] => {
                                      console.error('View provider errored:', error)
                                      return [asError(error)]
                                  }),
                                  // Add index to view to put response in right position of result views array below in scan operator
                                  map(view => ({ id: provider.id, view, index }))
                              )
                            : of(provider),
                    maxParallelQueries
                ),

                // Collect all responses to one result array
                scan(
                    (accumulator, current) => {
                        // Skip null step
                        if (current === null) {
                            return accumulator
                        }

                        const { index, ...payload } = current

                        accumulator[index] = payload

                        return accumulator
                    },
                    [null, ...providers.map(provider => ({ id: provider.id, view: undefined }))] as [
                        null,
                        undefined | { id: string; view: sourcegraph.View | null | undefined | ErrorLike }
                    ]
                )
            )
        ),
        // Filter all inappropriate values (nullish value and value with view: null)
        map(views => views.filter(allOf(isDefined, property('view', isNot(isExactly(null))))))
    )
}
