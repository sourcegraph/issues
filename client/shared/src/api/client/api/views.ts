import * as comlink from 'comlink'
import { isEqual, omit } from 'lodash'
import { combineLatest, ReplaySubject, Unsubscribable, ObservableInput, Subscription } from 'rxjs'
import { distinctUntilChanged, map, switchMap } from 'rxjs/operators'
import { PanelView, View } from 'sourcegraph'
import { ContributableViewContainer } from '../../protocol'
import { TextDocumentLocationProviderIDRegistry } from '../services/location'
import { PanelViewWithComponent, PanelViewProviderRegistry } from '../services/panelViews'
import { Location } from '@sourcegraph/extension-api-types'
import { MaybeLoadingResult } from '@sourcegraph/codeintellify'
import { ProxySubscribable } from '../../extension/api/common'
import { wrapRemoteObservable, ProxySubscription } from './common'
import { ViewService, ViewContexts } from '../services/viewService'
import { ExtensionHostAPI } from '../../extension/api/api'

/** @internal */
export interface PanelViewData extends Pick<PanelView, 'title' | 'content' | 'priority' | 'component'> {}

export interface PanelUpdater extends Unsubscribable, comlink.ProxyMarked {
    update(data: PanelViewData): void
}

/** @internal */
export interface ClientViewsAPI extends comlink.ProxyMarked {
    $registerDirectoryViewProvider(
        id: string,
        provider: comlink.Remote<
            ((context: ViewContexts[typeof ContributableViewContainer.Directory]) => ProxySubscribable<View | null>) &
                comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked

    $registerHomepageViewProvider(
        id: string,
        provider: comlink.Remote<
            ((context: ViewContexts[typeof ContributableViewContainer.Homepage]) => ProxySubscribable<View | null>) &
                comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked

    $registerInsightsPageViewProvider(
        id: string,
        provider: comlink.Remote<
            ((
                context: ViewContexts[typeof ContributableViewContainer.InsightsPage]
            ) => ProxySubscribable<View | null>) &
                comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked

    $registerGlobalPageViewProvider(
        id: string,
        provider: comlink.Remote<
            ((context: ViewContexts[typeof ContributableViewContainer.GlobalPage]) => ProxySubscribable<View | null>) &
                comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked
}

/** @internal */
export class ClientViews implements ClientViewsAPI {
    public readonly [comlink.proxyMarker] = true

    constructor(private viewService: ViewService) {}

    public $registerDirectoryViewProvider(
        id: string,
        provider: comlink.Remote<
            (
                context: ViewContexts[typeof ContributableViewContainer.Directory]
            ) => ProxySubscribable<View | null> & comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked {
        const subscription = new Subscription()
        subscription.add(
            this.viewService.register(id, ContributableViewContainer.Directory, context =>
                wrapRemoteObservable(provider(context), subscription)
            )
        )
        subscription.add(new ProxySubscription(provider))
        return comlink.proxy(subscription)
    }

    public $registerHomepageViewProvider(
        id: string,
        provider: comlink.Remote<
            (
                context: ViewContexts[typeof ContributableViewContainer.Homepage]
            ) => ProxySubscribable<View | null> & comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked {
        const subscription = new Subscription()
        subscription.add(
            this.viewService.register(id, ContributableViewContainer.Homepage, context =>
                wrapRemoteObservable(provider(context), subscription)
            )
        )
        subscription.add(new ProxySubscription(provider))
        return comlink.proxy(subscription)
    }

    public $registerInsightsPageViewProvider(
        id: string,
        provider: comlink.Remote<
            (
                context: ViewContexts[typeof ContributableViewContainer.InsightsPage]
            ) => ProxySubscribable<View | null> & comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked {
        const subscription = new Subscription()
        subscription.add(
            this.viewService.register(id, ContributableViewContainer.InsightsPage, context =>
                wrapRemoteObservable(provider(context), subscription)
            )
        )
        subscription.add(new ProxySubscription(provider))
        return comlink.proxy(subscription)
    }

    public $registerGlobalPageViewProvider(
        id: string,
        provider: comlink.Remote<
            (
                context: ViewContexts[typeof ContributableViewContainer.GlobalPage]
            ) => ProxySubscribable<View | null> & comlink.ProxyMarked
        >
    ): Unsubscribable & comlink.ProxyMarked {
        const subscription = new Subscription()
        subscription.add(
            this.viewService.register(id, ContributableViewContainer.GlobalPage, context =>
                wrapRemoteObservable(provider(context), subscription)
            )
        )
        subscription.add(new ProxySubscription(provider))
        return comlink.proxy(subscription)
    }
}
