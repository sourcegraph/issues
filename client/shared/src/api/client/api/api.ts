import { ClientContentAPI } from './content'
import { ClientContextAPI } from './context'
import { ClientViewsAPI } from './views'
import { ClientWindowsAPI } from './windows'
import { MainThreadAPI } from '../../contract'

/**
 * The API that is exposed from the client (main thread) to the extension host (worker)
 */
export interface ClientAPI extends MainThreadAPI {
    ping(): 'pong'

    context: ClientContextAPI
    windows: ClientWindowsAPI
    views: ClientViewsAPI
    content: ClientContentAPI
}
