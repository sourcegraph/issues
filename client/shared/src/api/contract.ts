import { SettingsCascade } from '../settings/settings'
import { SettingsEdit } from './client/services/settings'
import * as clientType from '@sourcegraph/extension-api-types'
import { Remote, ProxyMarked } from 'comlink'
import { Unsubscribable, TextDocument, Badged, DocumentHighlight } from 'sourcegraph'
import { ProxySubscribable } from './extension/api/common'
<<<<<<< HEAD:shared/src/api/contract.ts
import { TextDocumentPositionParams, ReferenceParams } from './protocol'
=======
import { TextDocumentPositionParameters } from './protocol'
>>>>>>> main:client/shared/src/api/contract.ts
import { MaybeLoadingResult } from '@sourcegraph/codeintellify'
import { HoverMerged } from './client/types/hover'
import { ConfiguredExtension } from '../extensions/extension'
import { ViewerData, ViewerId } from './viewerTypes'
import { TextDocumentIdentifier } from './client/types/textDocument'

/**
 * A text model is a text document and associated metadata.
 *
 * How does this relate to editors (in {@link ViewerService}? A model is the file, an editor is the
 * window that the file is shown in. Things like the content and language are properties of the
 * model; things like decorations and the selection ranges are properties of the editor.
 */
export interface TextDocumentData extends Pick<TextDocument, 'uri' | 'languageId' | 'text'> {}

/**
 * This is exposed from the extension host thread to the main thread
 * e.g. for communicating  direction "main -> ext host"
 * Note this API object lives in the extension host thread
 */
export interface FlatExtensionHostAPI {
    /**
     * Updates the settings exposed to extensions.
     */
    syncSettingsData: (data: Readonly<SettingsCascade<object>>) => void

    // Workspace
    addWorkspaceRoot: (root: clientType.WorkspaceRoot) => void
    getWorkspaceRoots: () => clientType.WorkspaceRoot[]
    removeWorkspaceRoot: (uri: string) => void
    syncVersionContext: (versionContext: string | undefined) => void

    // Search
    transformSearchQuery: (query: string) => ProxySubscribable<string>

    // Documents
    addTextDocumentIfNotExists: (textDocumentData: TextDocumentData) => void

    // Viewers

    addViewerIfNotExists: (viewerData: ViewerData) => ViewerId

    // Used for BlobPanel
    getActiveCodeEditorPosition: () => ProxySubscribable<TextDocumentPositionParams | null>
    getDecorations: (viewerId: ViewerId) => ProxySubscribable<clientType.TextDocumentDecoration[]>

    /**
     * Sets the selections for a CodeEditor.
     *
     * @param codeEditor The editor for which to set the selections.
     * @param selections The new selections to apply.
     * @throws if no editor exists with the given editor ID.
     * @throws if the editor ID is not a CodeEditor.
     */
    setEditorSelections(codeEditor: ViewerId, selections: clientType.Selection[]): void

    /**
     * Removes a viewer.
     * Also removes the corresponding model if no other viewer is referencing it.
     *
     * @param viewer The viewer to remove.
     */
    removeViewer(viewer: ViewerId): void

    // Languages
<<<<<<< HEAD:shared/src/api/contract.ts
    getHover: (parameters: TextDocumentPositionParams) => ProxySubscribable<MaybeLoadingResult<HoverMerged | null>>
    getDefinitions: (
        parameters: TextDocumentPositionParams
    ) => ProxySubscribable<MaybeLoadingResult<Badged<clientType.Location>[]>>
    getReferences: (parameters: ReferenceParams) => ProxySubscribable<MaybeLoadingResult<Badged<clientType.Location>[]>>
    hasReferenceProvider: (textDocument: TextDocumentIdentifier) => ProxySubscribable<boolean>
    getLocations: (
        id: string,
        parameters: TextDocumentPositionParams
    ) => ProxySubscribable<MaybeLoadingResult<Badged<clientType.Location>[]>>
    getDocumentHighlights: (
        parameters: TextDocumentPositionParams
    ) => ProxySubscribable<readonly clientType.DocumentHighlight[]>
=======
    getHover: (parameters: TextDocumentPositionParameters) => ProxySubscribable<MaybeLoadingResult<HoverMerged | null>>
    getDocumentHighlights: (parameters: TextDocumentPositionParameters) => ProxySubscribable<DocumentHighlight[]>
>>>>>>> main:client/shared/src/api/contract.ts
}

/**
 * This is exposed from the main thread to the extension host thread"
 * e.g. for communicating  direction "ext host -> main"
 * Note this API object lives in the main thread
 */
export interface MainThreadAPI {
    /**
     * Applies a settings update from extensions.
     */
    applySettingsEdit: (edit: SettingsEdit) => Promise<void>

    // Commands
    executeCommand: (command: string, args: any[]) => Promise<any>
    registerCommand: (
        name: string,
        command: Remote<((...args: any) => any) & ProxyMarked>
    ) => Unsubscribable & ProxyMarked

    getActiveExtensions: () => ProxySubscribable<ConfiguredExtension[]>
    getScriptURLForExtension: (bundleURL: string) => Promise<string>
}
