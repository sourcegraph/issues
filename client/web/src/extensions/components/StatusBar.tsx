import React, { useMemo } from 'react'
import { StatusBarItemWithKey } from '../../../../shared/src/api/client/api/codeEditor'
import { AbsoluteRepoFile } from '../../../../shared/src/util/url'
import { useObservable } from '../../../../shared/src/util/useObservable'
import classNames from 'classnames'
import { TextDocumentIdentifier } from '../../../../shared/src/api/client/types/textDocument'
import { Observable, timer } from 'rxjs'
import { useCarousel } from '../../components/useCarousel'
import MenuLeftIcon from 'mdi-react/MenuLeftIcon'
import MenuRightIcon from 'mdi-react/MenuRightIcon'
import { mapTo } from 'rxjs/operators'
import { Link } from 'react-router-dom'

interface StatusBarProps extends Omit<AbsoluteRepoFile, 'revision'> {
    getStatusBarItems: (parameters: TextDocumentIdentifier) => Observable<StatusBarItemWithKey[] | null>
}

export const StatusBar: React.FunctionComponent<StatusBarProps> = ({
    getStatusBarItems,
    repoName,
    filePath,
    commitID,
}) => {
    const uri = `git://${repoName}?${commitID}#${filePath}`

    const statusBarItems = useObservable(
        useMemo(
            () =>
                getStatusBarItems({
                    uri,
                }),
            [getStatusBarItems, uri]
        )
    )

    // Wait 3 seconds to show "no information from extensions avaiable"
    // to avoid UI jitter during initial extension activation.
    //
    // Restart timer whenever uri changes, since new extensions could be activated,
    // or could be fetching new data, so we include an unnecessary dependency.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    const hadTimeToLoad = useObservable(useMemo(() => timer(3000).pipe(mapTo(true)), [uri]))

    const {
        carouselReference,
        canScrollNegative,
        canScrollPositive,
        onNegativeClicked,
        onPositiveClicked,
    } = useCarousel({ direction: 'leftToRight' })

    return (
        <div className="status-bar w-100 border-top d-flex">
            {canScrollNegative && (
                <button type="button" className="btn btn-link status-bar__scroll border-0" onClick={onNegativeClicked}>
                    <MenuLeftIcon className="icon-inline" />
                </button>
            )}
            <div className="status-bar__items d-flex px-2" ref={carouselReference}>
                {statusBarItems
                    ? statusBarItems.map(statusBarItem => (
                          <StatusBarItem key={statusBarItem.key} statusBarItem={statusBarItem} />
                      ))
                    : hadTimeToLoad && (
                          <StatusBarItem
                              key="none-found"
                              statusBarItem={{
                                  key: 'none-found',
                                  text:
                                      'No information from extensions available. Find extensions in the Sourcegraph extension registry',
                              }}
                              component={
                                  <small className="text-muted">
                                      No information from extensions available.{' '}
                                      <Link to="/extensions">
                                          Find extensions in the Sourcegraph extension registry
                                      </Link>
                                  </small>
                              }
                          />
                      )}
            </div>
            {canScrollPositive && (
                <button type="button" className="btn btn-link status-bar__scroll border-0" onClick={onPositiveClicked}>
                    <MenuRightIcon className="icon-inline" />
                </button>
            )}
        </div>
    )
}

const StatusBarItem: React.FunctionComponent<{
    statusBarItem: StatusBarItemWithKey
    className?: string
    component?: JSX.Element
}> = ({ statusBarItem, className = 'status-bar', component }) => {
    const handleCommand = () => {}

    return (
        <div
            className={classNames(
                `${className}__item h-100 d-flex align-items-center px-1`,
                statusBarItem.tooltip && `${className}__item--tooltipped`
            )}
            data-tooltip={statusBarItem.tooltip}
            onClick={handleCommand}
        >
            {component || <small className={`${className}__text`}>{statusBarItem.text}</small>}
        </div>
    )
}
