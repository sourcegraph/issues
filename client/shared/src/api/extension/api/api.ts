import { ProxyMarked } from 'comlink'
import { InitData } from '../extensionHost'
<<<<<<< HEAD:shared/src/api/extension/api/api.ts
=======
import { ExtensionDocumentsAPI } from './documents'
import { ExtensionExtensionsAPI } from './extensions'
>>>>>>> main:client/shared/src/api/extension/api/api.ts
import { ExtensionWindowsAPI } from './windows'
import { FlatExtensionHostAPI } from '../../contract'

export type ExtensionHostAPIFactory = (initData: InitData) => ExtensionHostAPI

export interface ExtensionHostAPI extends ProxyMarked, FlatExtensionHostAPI {
    ping(): 'pong'

<<<<<<< HEAD:shared/src/api/extension/api/api.ts
=======
    documents: ExtensionDocumentsAPI
    extensions: ExtensionExtensionsAPI
>>>>>>> main:client/shared/src/api/extension/api/api.ts
    windows: ExtensionWindowsAPI
}
