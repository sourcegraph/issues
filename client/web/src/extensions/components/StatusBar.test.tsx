import React from 'react'
import { render } from '@testing-library/react'
import { StatusBar } from './StatusBar'
import { StatusBarItemWithKey } from '../../../../shared/src/api/client/api/codeEditor'
import { BehaviorSubject } from 'rxjs'

describe('StatusBar', () => {
    it('renders correctly', () => {
        const getStatusBarItems = () =>
            new BehaviorSubject<StatusBarItemWithKey[]>([
                { key: 'codecov', text: 'Coverage: 96%' },
                { key: 'code-owners', text: '2 code owners', tooltip: 'Code owners: @felixbecker, @beyang' },
            ]).asObservable()

        expect(
            render(<StatusBar getStatusBarItems={getStatusBarItems} repoName="test" commitID="test" filePath="test" />)
                .baseElement
        ).toMatchSnapshot()
    })
})
