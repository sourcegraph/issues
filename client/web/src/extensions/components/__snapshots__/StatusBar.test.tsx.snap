// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBar renders correctly 1`] = `
<body>
  <div>
    <div
      class="status-bar w-100 border-top px-2 d-flex"
    >
      <div
        class="status-bar__item h-100 d-flex align-items-center px-1"
      >
        <small
          class="status-bar__text"
        >
          Coverage: 96%
        </small>
      </div>
      <div
        class="status-bar__item h-100 d-flex align-items-center px-1 status-bar__item--tooltipped"
        data-tooltip="Code owners: @felixbecker, @beyang"
      >
        <small
          class="status-bar__text"
        >
          2 code owners
        </small>
      </div>
    </div>
  </div>
</body>
`;
